R/integrate_sces.R

```diff
@@ -0,0 +1,180 @@
+#'  integration method. The final SCE object will contain an additional reducedDim
+#'  entitled `{integration_method}_PCA`. All original SCE assays are retained.
+#'  `fastMNN` integration uses all default setting, but additional parameters can be supplied.
+#'  `harmony` integration uses existing PCs, but additional parameters can be supplied.
+#'
+#' @param merged_sce A merged SCE object as prepared by `scpcaTools::merge_sce_list()`.
+#' @param integration_method The integration method to use. One of `fastMNN` or
+#'  `harmony` (case-sensitive)
+#' @param batch_column The column in the merged SCE object indicating batches
+#' @param covariate_cols A vector of additional columns in the merged SCE to
+#'   consider as covariates during integration. Currently, this is used only by
+#'   `harmony`.
+#' @param return_corrected_expression A boolean indicating whether corrected expression
+#'   values determined by the given integration method should be included in the
+#'   integrated SCE object. Note that `harmony` does not calculate corrected expression,
+#'   so this argument is ignored for this integration method. Default is `TRUE`.
+#' @param seed Random seed to set for integration. This is only set if the value is not `NULL`.
+#' @param ... Any additional parameters to be passed to the given integration method
+#'
+#' @return An updated merged SCE object containing a new reduced dimension named
+#'   `{integration_method}_PCA` containing the corrected PCs. In the case of `fastMNN`
+#'   integration, the SCE also includes a corrected expression assay (`fastMNN_corrected`)
+#'
+#' @import SingleCellExperiment
+#'
+#' @export
+integrate_sces <- function(merged_sce,
+                           integration_method = c("fastMNN", "harmony"),
+                           batch_column = "sample",
+                           covariate_cols = c(),
+                           return_corrected_expression = TRUE,
+                           seed = NULL,
+                           ...) {
+
+  # Set seed
+  set.seed(seed)
+
+  # make sure that input is a SingleCellExperiment
+  if(!is(merged_sce, "SingleCellExperiment")){
+    stop("The `merged_sce` must be a SingleCellExperiment object created with `scpcaTools::merge_sce_list().`")
+  }
+
+
+  # Check integration_method
+  integration_method <- match.arg(integration_method)
+
+  # Check batch_column
+  if (!(batch_column %in% names(colData(merged_sce)))) {
+    stop("The provided `batch_column` column must be in the `merged_sce` colData.")
+  }
+  # Ensure there are no NAs
+  batches <- colData(merged_sce)[[batch_column]]
+  if( sum(is.na(batches)) > 0 ) {
+    stop("`NA` values are not allowed in the `batch_column`.")
+  }
+  # Ensure >=2 batches
+  if( length(unique(batches)) < 2 ) {
+    stop("At least two batches are required for integration.")
+  }
+
+
+  # Perform integration with specified method
+  if (integration_method == "fastMNN") {
+
+    if (length(covariate_cols) >= 1) {
```

Comment:
I don't know why, but...
```suggestion
    if (length(covariate_cols) > 0) {
```

I mean, we could even do:
```
if (length(covariate_cols)) {
```
which would be more pythonic, but this is R. (Really Pythonic would actually be `if(covariate_cols)`, but that won't work in R)